File: chess-app/src/engine/pieces/Piece.js
export default class Piece {
    constructor(type, color) {
        this.type = type;
        this.color = color;
        this.hasMoved = false;
    }

    getPseudoLegalMoves(board, fromPos) {
        throw new Error("getPseudoLegalMoves() must be implemented in subclass");
    }

    _getSlidingMoves(directions, board, fromPos) {
        const moves = [];
        directions.forEach((dir) => {
            for(let i=1; i<8; i++) {
                let newPos = fromPos.map((value, index) => value + i * dir[index])
                if (!board.isInsideBoard(newPos)){
                    break;
                }
                let pieceAtNewPos = board.getPieceAt(newPos);
                if (pieceAtNewPos !==null && pieceAtNewPos.color === this.color) {
                    break;
                } else if (pieceAtNewPos !==null && pieceAtNewPos.color !== this.color) {
                    moves.push(newPos);
                    break;
                } else {
                    moves.push(newPos)
                }
            }   
        })
        return moves
    }

    _getSteppingMoves(offsets, board, fromPos) {
        const moves = []
        offsets.forEach((offset)=>{
            let newPos = fromPos.map((value, index) => value + offset[index]);
            if (board.isInsideBoard(newPos)){
                let pieceAtNewPos = board.getPieceAt(newPos)
                if (pieceAtNewPos === null || pieceAtNewPos.color !== this.color){
                    moves.push(newPos)
                }
            }
        })
        return moves
    }

}

File: chess/src/main/java/com/app/chess/Pieces/Piece.java
package com.app.chess.Pieces;
import com.app.chess.*;

import java.util.List;
import java.util.ArrayList;

public class Piece {
    public PieceType type;
    public Player color;
    public boolean hasMoved;

    public Piece(PieceType ptype, Player pcolor) {
        type = ptype;
        color = pcolor;
        hasMoved = false;
    }

    public Piece(PieceType ptype) {
        type = ptype;
        color = null;
        hasMoved = false;
    }

    public Piece() {
        type = null;
        color = null;
        hasMoved = false;
    }

    public List<int []> getPsuedoLegalMoves(Board board, int[] position) {
        // This method should be overridden in each specific piece class
        return List.of();
    }

    public List<int []> getPsuedoLegalMoves(Board board, int[] position, boolean attack) {
        // This method should be overridden in each specific piece class
        return List.of();
    }

    public Piece deepCopy() {
        // This method should be overridden in each specific piece class
        return new Piece();
    }

    public Piece setPlayer(Player color) {
        this.color = color;
        return this;
    }

    protected List<int []> getSlidingMoves(List<int []> directions, Board board, int[] fromPos) {
        List <int []> legalMoves = new ArrayList<>();
        int fR = fromPos[0];
        int fC = fromPos[1];
        for (int [] dir : directions) {
            while (true) {
                fR += dir[0];
                fC += dir[1];
                int[] nextPos = {fR, fC};
                if (!board.isInsideBoard(nextPos)) {
                    break;
                }
                Piece pieceAtNext = board.getPieceAt(nextPos);
                if (pieceAtNext == null) {
                    legalMoves.add(nextPos);
                } else {
                    if (pieceAtNext.color != this.color) {
                        legalMoves.add(nextPos);
                    }
                    break;
                }
            }
        }
        return legalMoves;
    }

    protected List<int []> getSteppingMoves(List<int []> directions, Board board, int[] fromPos) {
        List <int []> legalMoves = new ArrayList<>();
        int fR = fromPos[0];
        int fC = fromPos[1];
        for (int [] dir : directions) {
            int[] nextPos = {fR + dir[0], fC + dir[1]};
            if (!board.isInsideBoard(nextPos)) {
                continue;
            }
            Piece pieceAtNext = board.getPieceAt(nextPos);
            if (pieceAtNext == null || pieceAtNext.color != this.color) {
                legalMoves.add(nextPos);
            }
        }
        return legalMoves;
    }
}
